### DISCUSSION TITLE
Switch is not evaluating after track end

### DISCUSSION THREAD

**Discussion by jdbuys** [savonet/liquidsoap#3287] [User Support]
**Describe the bug**
When I have a script where one source uses the crossfade operator, then the switch operator does not re-evaluate when the track ends, but if I remove the crossfade then I can see it evaluates after every track.

**To Reproduce**
```
#!/usr/bin/liquidsoap
# Log dir
log.file.path.set("/tmp/basic-radio.log")
log.stdout.set(true)
log.level.set(4)
log.file.set(true)
settings.server.telnet.set(false)
settings.server.socket.set(true)
settings.server.socket.path.set("lsoap.sock")
settings.server.socket.permissions.set(511)
settings.server.timeout.set(20.)
settings.log.file.perms.set(511)
activesource=interactive.int("activesource", 3)


def is_active(id) =
 log.info("Cheking if source is active")
 if (id == activesource()) then
    true
  else
   false
 end
end

# Music
myplaylist = crossfade(playlist("/tmp/playlist.pls"))
#myplaylist = playlist("/tmp/playlist.pls")
source2= input.http(id="1","http://127.0.0.1/idTest2",start=true,max_buffer=25., timeout=15.,poll_delay=10.)
source3= input.http(id="2","http://127.0.0.1/idTest3",start=false,max_buffer=25., timeout=15.,poll_delay=10.)

radio = mksafe(switch(all_predicates=true, track_sensitive=true, [({is_active(1)}, source2), ({is_active(2)}, source3), ({is_active(3)}, myplaylist)]))

# Stream it out
output.icecast(%mp3,
  host = "192.168.10.123", port = 80,
  password = "hackme", mount = "idTest5",user="source",
  radio)
```

**Expected behavior**
I expected the switch to evaluate after every track ends. 

**Version details**

- OS: Ubuntu 22.04
- Version 2.1.4 and 2.2.0

**Install method**
Built from source

**Comment by toots** [comment]
Hi @jdbuys. Currently, `crossfade` might indeed remove track marks. This is because there is no clear notion of track marks when overlapping track start and stop.

There are several ways to fix this but first: how would you expect your track sensitive switch to work in the presence of a crossfade where two tracks are being played at the same time?

**Reply by jdbuys** [reply]
Hi @toots 
Good question. I guess it would be great if the currently playing track fades out and then the switch happens.

**Reply by jdbuys** [reply]
What I am trying to do is: Based on some predicate switch to the Icecast source when the playlist track ends, and then later switch back to the playlist.

**Reply by toots** [reply]
Ok. I think the best solution here is to have a custom crossfade transition.

When the transition is computed, if one of the icecast source is available, you could just do a `sequence([old, new])`. This should actually keep the track marks and everything should work as intended.

**Comment by jdbuys** [comment]
Thanks @toots , I think I have a working solution.

**Reply by toots** [reply]
Great!
